analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ParameterValidationInAsyncShouldBeWrapped.CSharp10.cs

```diff
@@ -0,0 +1,97 @@
+﻿using System;
+using System.IO;
+using System.Threading.Tasks;
+
+namespace Tests.Diagnostics
+{
+    public class InvalidCases
+    {
+        public static async Task<string> FooAsync(string something) // FN
+        {
+            ArgumentNullException.ThrowIfNull(something); // FN sec
+
+            await Task.Delay(1);
+            return something + "foo";
+        }
+
+        // See https://github.com/SonarSource/sonar-dotnet/issues/2665
+        public async void OnSomeEvent(object sender, EventArgs args) // Compliant
+        {
+            ArgumentNullException.ThrowIfNull(sender);
+
+            await Task.Yield();
+        }
+
+        public void Foo(object sender, EventArgs args) { }
+    }
+
+    public class ValidCases
+    {
+        public static Task FooAsync(string something)
+        {
+            ArgumentNullException.ThrowIfNull(something);
+
+            return FooInternalAsync(something);
+        }
+
+        private static async Task<string> FooInternalAsync(string something)
+        {
+            await Task.Delay(1);
+            return something + "foo";
+        }
+
+        public async Task DoAsync() // Compliant - no args check
+        {
+            await Task.Delay(0);
+        }
+
+        public async Task FooAsync(int age) // Compliant - the exception doesn't derive from ArgumentException
```

Comment:
```suggestion
        public async Task FooAsync(int age) // Compliant - the exception doesn't derive from ArgumentNullException
```

Reply:
Does it make sense to add a test case with an exception derived from `ArgumentNullException`? I don't see any use case for such a derived class, but the comment suggests, that you had something like this in mind as a case.